### ISSUE TITLE
Cortex Analyzers "No output" error

### ISSUE THREAD

**Issue by b00lpy** [TheHive-Project/Cortex#211]
# Cortex Analyzers unknown error

### Request Type
Bug

### Work Environment

| Question              | Answer
|---------------------------|--------------------
| OS version (server)       |  Ubuntu Server 18.04 LTS
| OS version (client)       | Ubuntu 18.04/Win 10
| Cortex version / git hash   | 3.0.0-RC3
| Package Type              |  Docker
| Browser type & version    | Firefox/Chrome

### Problem Description
When i start a job with a Cortex Analyzer, in almost 1/2 of the enabled analyzer I got an unknown error which says there's no output. Don't know what can be the cause.
Tried both enable and disable proxy but it keeps doing it

### Steps to Reproduce
1. Enable an analyzer (my case is URLhaus)
2. Start a job with that
3. Wait report

### Complementary information
Screenshot:  

![immagine](https://user-images.githubusercontent.com/50588190/61703086-282c9280-ad41-11e9-87ca-e63329a30126.png)

**Comment by github-pba** [comment]
I had some issues, too. Most of the time it helped to change the Python interpreter from version 2 to 3.

You could either change your default pyhton engine for the entire machine, or replace the very first line of each responder not working with the call of python 3.

```
#!/usr/bin/env python
```
becomes
```
#!/usr/bin/env python3
```

**Comment by github-pba** [comment]
To verify python does not throw any errors, you could call an analyzer from the command line interface, such as

```
echo '{"data":"http://89.40.15.153/apache2","dataType":"url","tlp":0}' | python3 URLhaus_analyzer.py
```

**Comment by b00lpy** [comment]
Sorry for my late response. 
By the way I tried build a custom tool and testi it with the command line script you wrote:
`echo '{"data":"http://89.40.15.153/apache2","dataType":"url","tlp":0}' | python3 URLhaus_analyzer.py
`.

The results are quite bad.
My tool works, I built a detailed logger daemon and inside logs I can see all steps perfomed until the creation of "Report" object and the output dictionary . No warnings, no errors, no exception.

But also in my tool ( and also HashDD for example ) i keep seeing this output. Cortex logs simply says it fails and it's been two days I keep modifying the script without success.

**Comment by github-pba** [comment]
Can you publish your script somewhere, so I could download it and run it on my system?

**Comment by github-pba** [comment]
... ah, and what happens if you run the code snippet I have sent you:

`echo '{"data":"http://89.40.15.153/apache2","dataType":"url","tlp":0}' | python3 URLhaus_analyzer.py`

**Comment by b00lpy** [comment]
Actually inside the script there are some private business configurations, maybe I'll take some time to anonymize it and i'll upload online.
By the way the problem ain't the script itself, if I run them via command line everything works. 

URLhaus works fine, also through cortex.

Now I tried to edit some steps and copying the structure of the `URLhaus` one. Now I got this:
```
{
  "errorMessage": "JsResultException(errors:List((,List(JsonValidationError(List(error.expected.jsobject),WrappedArray())))))",
  "input": "{\"data\":\"bdd5f57ed46a777142794cf1c5d25402\",\"dataType\":\"hash\",\"tlp\":2,\"message\":\"\",\"parameters\":{},\"config\":{\"proxy_https\":null,\"cacerts\":null,\"max_pap\":2,\"jobTimeout\":30,\"check_tlp\":true,\"proxy_http\":null,\"max_tlp\":2,\"auto_extract_artifacts\":true,\"jobCache\":10,\"check_pap\":true},\"pap\":2}",
  "success": false
}
```

**Comment by github-pba** [comment]
OK, you say

> URLhaus works fine, also through cortex.

You first reported you cannot get URLhaus make work. This issue seems to be solved, right?
Do the other analyzers work, too?

**Comment by github-pba** [comment]
Check out [this](https://github.com/mesosphere/marathon/issues/887): looks similar.

And did you recognize that a parameter is missing?
![ss001](https://user-images.githubusercontent.com/50981940/62937648-89d69e80-bdcd-11e9-9841-c95ab0d99855.png)

**Comment by b00lpy** [comment]
> OK, you say
> 
> > URLhaus works fine, also through cortex.
> 
> You first reported you cannot get URLhaus to work. This issue seems to be solved, right?
> Do the other analyzers work, too?

You are right, I reinstalled Cortex inside another machine and maybe they fixed this error, but I got the same in other tools. I'm still using a the container version, so maybe this can be the problem.

For the other thing, I already found that Issue for `marathon` but didn't helped so much. I almost "copied" the `URLhaus.py` code and seems that the error is inside function `def summary(self):` if the analyzer class. But I don't know why because the return value it's the same, a simple JSON.

Here's my summary function:
```
    def summary(self, raw):
        taxo = self.report_content["summary"]

        taxonomies = []

        taxonomies.append(self.build_taxonomy(
            taxo["predicate"],
            "Custom_Analysis_Tool",
            taxo["value"],
            taxo["level"]
        ))
        logger("Returning taxonomies")
        try:
            logger(taxonomies.__str__())
        except:
            pass
        return {"taxonomies": taxonomies}
```

**Comment by github-pba** [comment]
Since the original problem is solved, you should probably close this issue and open a new one with a new problem description. I have no experience with writing analyzers, so I cannot help any further. Sorry.

The subject and the long discussion between us will probably make nobody read through this thread and pick up where I have to leave.

Please be aware, if you open an issue with a subject like "trouble getting analyzer to work", you surely have to provide the source code, so experienced analyzer developers can look at it.

Good luck and best wishes.

**Comment by b00lpy** [comment]
My problem is that I have the same error both in my custom and in some of the "official", but I will surely do what you suggested.

Thank you again for your support! ;)

**Comment by github-pba** [comment]
b00lpy, you're welcome.

When you have the same error in official analyzers, you should focus on these, as they can best be supported. 

What I do not understand is *what* problem you still have. When we started, URLhaus said `no output`. Then URLhaus works, and you start to write your own analyzer, which results in the `"errorMessage": "JsResultException ...` problem. And now you say that "your and other official analyzers have the same problem.

You really can name official analyzers which ouput `"errorMessage": "JsResultException ..`???

**Comment by github-pba** [comment]
Ah, one last thing: in your new posting, place the output of the two commands `pip show cortexutils` and `pip3 show cortexutils`

**Comment by github-pba** [comment]
I found a last thing:

in `/etc/cortex/logback.xml`, set `debug="true"` in line 2. Maybe this gives you more information in `/var/log/cortex/application.log`

**Comment by nadouani** [comment]
Hello, this looks hard to follow, I don't know how does your analyzer get a `JsResultException ` which is something that Cortex can return.

Is your analyzer calling Cortex APIs?

Can you describe what does your analyzer do?

**Comment by nadouani** [comment]
Does the problem occur for "official" analyzers we provide or is it part of your own custom analyzers?

**Comment by b00lpy** [comment]
Answering to @github-pba , cortexutils is updated to 2.0.0 so I think it's ok.

@nadouani probably this long tread is confusing.
My actual situation is:
Some "official" analyzers gave me this `no output` error, don't know why this kind of error seems to reappear sometimes. I opened this issue about URLhaus, next it fixed itself and now I have that problem again.
![image](https://user-images.githubusercontent.com/50588190/63433288-31ce1680-c423-11e9-8d5f-8810abadc5c1.png)

And talking about my custom analyzer, I've got the `JsResultException` error. It's a simple script which checks for the observable value inside local blacklists. The script itself works, I tried to use the VirusTotal analyzer as template and edit it to fit my script and here's the error.
I post only the output:
```
{
  "errorMessage": "JsResultException(errors:List((,List(JsonValidationError(List(error.expected.jsobject),WrappedArray())))))",
  "input": "{\"data\":\"www.test.noe\",\"dataType\":\"url\",\"tlp\":2,\"message\":\"\",\"parameters\":{},\"config\":{\"proxy_https\":null,\"cacerts\":null,\"max_pap\":2,\"jobTimeout\":30,\"check_tlp\":true,\"proxy_http\":null,\"max_tlp\":2,\"auto_extract_artifacts\":true,\"jobCache\":10,\"check_pap\":true},\"pap\":2}",
  "success": false
}
```

**Comment by b00lpy** [comment]
### CUSTOM ANALYZER 
EDIT: While inside Cortex GUI i see that `JsResultException` in Cortex logs i see this:

![image](https://user-images.githubusercontent.com/50588190/63606585-b4e09f80-c5d0-11e9-9fd5-fdf0600937db.png)

disclaimer: IP is just random typing
